src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs

```diff
@@ -495,35 +490,6 @@ private void VerifyPathsForEntryType(string filePath, string? linkTargetPath, bo
                 throw new IOException(SR.Format(SR.IO_AlreadyExists_Name, filePath));
             }
             File.Delete(filePath);
-
-            if (EntryType is TarEntryType.SymbolicLink or TarEntryType.HardLink)
```

Comment:
I looked at the tar change history. I remember we introduced this removed code section as a preventative measure. Here is the commit: https://github.com/dotnet/runtime/pull/67883/commits/5d815778eb5ea9c4a2fda774472c40b824c0bf02

The commit name is "Disallow extracting individual links". The limitation prevents someone from using the TarFile APIs  to point to a file that does not yet exist.

I'd be happy to revisit this. The removed code has two parts:

A) Prevent extracting a symlink whose target does not yet exist.
B) Prevent extracting a hardlink whose target does not yet exist, or attempting to create a hardlink that points to a directory.

@GrabYourPitchforks - Do you think these checks could be removed so TarFile is not so restricted? Notice @tmds added a new test to confirm that a symlink can be extracted when the target does not yet exist.

Reply:
> @GrabYourPitchforks - Do you think these checks could be removed so TarFile is not so restricted? Notice @tmds added a new test to confirm that a symlink can be extracted when the target does not yet exist.

Tar writers don't guarantee that symbolic link targets are in the archive before the symbolic links, which is what this code was assuming.

> B) Prevent extracting a hardlink whose target does not yet exist, or attempting to create a hardlink that points to a directory.

hardlinks can never be created to targets that don't exist, or to a directory.

Reply:
Note that this does **not** change the restriction that link targets point in the destination directory.

Reply:
It may be fine to drop the "target does not yet exist" validation for symbolic/hard links. `File.CreateSymbolicLink` does not enforce that the target exists so I think is fine to align the behavior here.

Reply:
Yes, they can be dropped. I'll include a comment explaining why.